fix(metrics): Replace kube-rbac-proxy with built-in auth (#573)

fyzanshaik opened the pull request:
…ics point

**Description**

This PR fixes #

**Notes for Reviewers**


**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 

<!--
Thank you for contributing to Meshery! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

fyzanshaik commented:
This PR addresses the deprecation of the `gcr.io/kubebuilder/kube-rbac-proxy` image by migrating the `meshery-operator`'s metrics endpoint protection mechanism. Instead of relying on the external `kube-rbac-proxy` sidecar container, this change implements the recommended upstream approach using the built-in authentication and authorization features provided by `controller-runtime`.

**Changes Implemented:**

1.  **`main.go` Modifications:**
    *   Imports added for `crypto/tls`, `metricsserver`, and `metrics/filters`.
    *   Introduced command-line flags:
        *   `--metrics-bind-address` (defaults to `:8443`): Sets the metrics server address.
        *   `--secure-metrics` (defaults to `true`): Enables/disables secure serving (HTTPS and AuthN/AuthZ) for metrics.
        *   `--enable-http2` (defaults to `false`): Controls HTTP/2 enablement for metrics and webhooks (disabled by default for security).
    *   Configured `metricsserver.Options`:
        *   Sets `BindAddress` based on the flag.
        *   Sets `SecureServing` based on the flag (defaults to `true`).
        *   Configures `TLSOpts` (disabling HTTP/2 by default).
        *   Conditionally sets `FilterProvider: filters.WithAuthenticationAndAuthorization` when `secureMetrics` is true. This enables the built-in TokenReview/SubjectAccessReview checks.
    *   Updated Webhook server configuration to use the same `TLSOpts`.
    *   Passed the configured `metricsserver.Options` to `ctrl.NewManager`.
    *   Ran `go mod tidy` to update `go.mod` and `go.sum` with new dependencies (like `k8s.io/apiserver` pulled in by the filters).

2.  **Required Configuration Changes (Not included in `main.go` changes, must be done separately):**
    *   **Deployment Template (`install/kubernetes/helm/meshery-operator/templates/deployment.yaml`):** The container definition for `kube-rbac-proxy` **must be removed**.
    *   **Helm Values (`install/kubernetes/helm/meshery-operator/values.yaml`):** The entire `kubeRbac:` section **must be removed**. Consider adding new values to control the `--secure-metrics` flag if needed.
    *   **Service Template (`install/kubernetes/helm/meshery-operator/templates/metrics_service.yaml` or similar):** The Service definition targeting the metrics endpoint **must be updated** to target port `8443` and likely use the name `https` instead of `http`.
    *   **RBAC (`install/kubernetes/helm/meshery-operator/templates/` or `rbac/`):**
        *   New `ClusterRole` definitions are required:
            *   One granting the operator's ServiceAccount permissions for `authentication.k8s.io/tokenreviews` (verb: `create`) and `authorization.k8s.io/subjectaccessreviews` (verb: `create`).
            *   One (e.g., `metrics-reader`) granting permissions for nonResourceURLs (`/metrics`) with verb `get`.
        *   New `ClusterRoleBinding` definitions are required:
            *   One binding the operator's ServiceAccount to the role granting `tokenreviews`/`subjectaccessreviews`.
            *   One binding the entity responsible for scraping metrics (e.g., Prometheus ServiceAccount) to the `metrics-reader` role. (Examples are in the Kubebuilder migration guide linked in #573).

**New Behavior & Configuration:**

*   By default, the operator now serves metrics securely over HTTPS on port `:8443`.
*   Accessing the `/metrics` endpoint requires a valid Kubernetes bearer token from an entity authorized by the RBAC rules (specifically, bound to the `metrics-reader` role or equivalent).
*   The operator logs whether metrics are being served securely and if HTTP/2 is disabled.
*   TLS uses auto-generated self-signed certificates by default (suitable for testing, but **should be replaced with proper certificates** via cert-manager or secrets for production deployments).

aabidsofi19 commented:
@fyzanshaik  regarding the #2 set of changes  i assume they are yet to be added in this pr

fyzanshaik commented:
Removes the dependency on the deprecated gcr.io/kubebuilder/kube-rbac-proxy
image by migrating metrics endpoint protection to use the built-in
authentication and authorization features of controller-runtime.

Key changes include:

- Modified cmd/main.go:
    - Enabled secure metrics serving (:8443/https) by default.
    - Configured metrics server using `metricsserver.Options`.
    - Added `filters.WithAuthenticationAndAuthorization` provider.
    - Added flags (`--secure-metrics`, `--enable-http2`) for configuration.
    - Updated `go.mod`/`go.sum` via `go mod tidy`.
- Removed Kustomize Proxy Patch:
    - Deleted `config/default/manager_auth_proxy_patch.yaml`.
    - Removed reference from `config/default/kustomization.yaml`.
- Removed Proxy RBAC & Service:
    - Deleted `config/rbac/auth_proxy_service.yaml`.
    - Deleted `config/rbac/auth_proxy_role.yaml`.
    - Deleted `config/rbac/auth_proxy_role_binding.yaml`.
    - Removed references from `config/rbac/kustomization.yaml`.
- Updated Core & Metrics RBAC:
    - Renamed `auth_proxy_client_clusterrole.yaml` to `metrics_reader_role.yaml`.
    - Updated `config/rbac/kustomization.yaml` for `metrics_reader_role.yaml`.
    - Added explicit `tokenreviews` & `subjectaccessreviews` permissions
      to `config/rbac/controller_role.yaml`.
- Updated Metrics Service:
    - Modified `bundle/.../meshery-operator-metrics-service_v1_service.yaml`
      to target port 8443 and use port name 'https'.
- Updated Manager Deployment:
    - Modified `config/manager/manager.yaml` to update metrics port definition
      to 8443/https.
- Updated ServiceMonitor:
    - Modified `config/prometheus/monitor.yaml` to use `https` scheme,
      target port 'https', and add `bearerTokenFile` / `tlsConfig`.
- Updated OLM Bundle:
    - Modified `bundle/.../meshery-operator.clusterserviceversion.yaml`:
        - Removed `kube-rbac-proxy` container from deployment spec.
        - Updated `manager` container args (removed `--metrics-addr`) and
          ports (updated metrics port to 8443/https).

This aligns the operator with current Kubebuilder best practices for
securing metrics endpoints.

fyzanshaik commented:
With the changes made to:

    main.go
    config/rbac/controller_role.yaml
    config/rbac/kustomization.yaml (removing proxy RBAC, renaming/keeping metrics reader)
    config/default/kustomization.yaml (removing proxy patch reference)
    bundle/0.0.1/manifests/meshery-operator-metrics-service_v1_service.yaml (updating service ports)
    config/manager/manager.yaml (updating container port definition)
    config/prometheus/monitor.yaml (updating scheme, port, adding auth/tlsConfig)
    bundle/0.0.1/manifests/meshery-operator.clusterserviceversion.yaml (updating deployment spec)
    And deleting the various proxy-related files (manager_auth_proxy_patch.yaml, auth_proxy_service.yaml, auth_proxy_role.yaml, auth_proxy_role_binding.yaml)

We should now have completed all the necessary changes within the meshery-operator repository to address issue #573 by migrating to the built-in controller-runtime metrics authentication.

The next focus would be on the meshery repository to handle any installation or metrics scraping configuration adjustments needed there.

aabidsofi19 commented on bundle/0.0.1/manifests/meshery-operator-metrics-service_v1_service.yaml:
```diff
@@ -1,18 +1,18 @@
-    targetPort: https
+    - name: https
+      port: 8443
+      targetPort: 8443 # <-- CHANGE THIS from 'https' to '8443'
```
do we need to update any client to use https

aabidsofi19 commented on bundle/0.0.1/manifests/meshery-operator.clusterserviceversion.yaml:
```diff
@@ -500,36 +500,23 @@ FnZVJlYWR5ccllPAAAAABJRU5ErkJggg=="
                   name: server
-                - containerPort: 8080
-                  hostPort: 8080
+                - containerPort: 8443
```
do we need to update meshery to use this port

aabidsofi19 commented:
were you able to test these changes locally ?

leecalcote commented:
@fyzanshaik, your repetitive comments make review arduous. Inclusion of these specifics is healthy. Doing so repetitively means that others have to wade through your LLM-generated minutiae. Please spare us.

fyzanshaik commented:
> @fyzanshaik, your repetitive comments make review arduous. Inclusion of these specifics is healthy. Doing so repetitively means that others have to wade through your LLM-generated minutiae. Please spare us.
> 
> It smells of AI 🤮 and laziness to sift through it on your part, passing along the 🤮 for others to deal with.

it is AI. I just tried to solve the issue on my own. I did inform Aabid about it.
